### ISSUE TITLE
Surprising `mpirun -np 1 which python`

### ISSUE THREAD

**Issue by paugier** [xonsh/xonsh#3617]
I'm trying to understand how to use xonsh. This is particularly disturbing and looks like a bug!

```
pierre@kth ~ $ python --version                                                 
Python 3.8.3
pierre@kth ~ $ mpirun -np 1 python --version                                    
Python 2.7.18rc1
pierre@kth ~ $ which python                                                     
/home/pierre/.pyenv/shims/python
pierre@kth ~ $ mpirun -np 1 which python                                        
/usr/bin/python
pierre@kth ~ $ echo $PATH                                                       
/home/pierre/miniconda3/condabin/app:/home/pierre/opt/julia-1.0.2/bin/:/home/pierre/.pyenv/shims:/home/pierre/.pyenv/bin:/home/pierre/.local/bin/:/home/pierre/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
pierre@kth ~ $ mpirun -np 1 echo $PATH                                          
/home/pierre/miniconda3/condabin/app:/home/pierre/opt/julia-1.0.2/bin/:/home/pierre/.pyenv/shims:/home/pierre/.pyenv/bin:/home/pierre/.local/bin/:/home/pierre/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
```

## xonfig

(xonsh from master)

<details>

```
$ xonfig
+------------------+-----------------+
| xonsh            | 0.9.18          |
| Python           | 3.8.3           |
| PLY              | 3.11            |
| have readline    | True            |
| prompt toolkit   | 3.0.5           |
| shell type       | prompt_toolkit  |
| pygments         | 2.6.1           |
| on posix         | True            |
| on linux         | True            |
| distro           | ubuntu          |
| on darwin        | False           |
| on windows       | False           |
| on cygwin        | False           |
| on msys2         | False           |
| is superuser     | False           |
| default encoding | utf-8           |
| xonsh encoding   | utf-8           |
| encoding errors  | surrogateescape |
+------------------+-----------------+
```

</details>

**Comment by paugier** [comment]
Of course,
```
pierre@kth ~ $ bash
pierre@kth:~$ mpirun -np 1 which python
/home/pierre/.pyenv/shims/python
```

**Comment by laloch** [comment]
Hi @paugier, I believe `mpirun -np 1 echo $PATH` does not really do what you think it does. The `$PATH` variable gets expanded **before** `mpirun` is called.

**Comment by laloch** [comment]
Hmm, thinking about it, `/bin/echo` probably cannot expand variables either, so you would probably want to run `mpirun -np 1 python -c "import os; print(os.environ['PATH'])"` in order to find out what the `PATH` value really is within `mpirun`.

**Comment by paugier** [comment]
Of course! Thank you @laloch 
```
pierre@gre ~ $ echo $PATH                                                       
/home/pierre/.cargo/bin:/data0/opt/miniconda3/condabin/app:/data0/opt/miniconda3/condabin:/home/pierre/.pyenv/shims:/home/pierre/.pyenv/bin:/home/pierre/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
pierre@gre ~ $ mpirun -np 1 python -c "import os; print(os.environ['PATH'])"    
/usr/bin:/usr/bin:/home/pierre/.cargo/bin:/data0/opt/miniconda3/condabin/app:/data0/opt/miniconda3/condabin:/home/pierre/.pyenv/shims:/home/pierre/.pyenv/bin:/home/pierre/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
```
So PATH is modified by mpirun when launched from xonsh (and not when launched from bash).

**Comment by laloch** [comment]
Or maybe Xonsh fails to prepare the environment for the child process properly...
Is there a difference between:
```
$ echo $PATH
$ python -c "import os; print(os.environ['PATH'])"
$ mpirun -np 1 python -c "import os; print(os.environ['PATH'])"
```
?

**Comment by paugier** [comment]
```
$ echo $PATH                                                       
/home/pierre/.cargo/bin:/data0/opt/miniconda3/condabin/app:/data0/opt/miniconda3/condabin:/home/pierre/.pyenv/shims:/home/pierre/.pyenv/bin:/home/pierre/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
$ python -c "import os; print(os.environ['PATH'])"                 
/home/pierre/.pyenv/versions/3.8.2/bin:/home/pierre/.pyenv/libexec:/home/pierre/.pyenv/plugins/python-build/bin:/home/pierre/.cargo/bin:/data0/opt/miniconda3/condabin/app:/data0/opt/miniconda3/condabin:/home/pierre/.pyenv/shims:/home/pierre/.pyenv/bin:/home/pierre/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
$ mpirun -np 1 python -c "import os; print(os.environ['PATH'])"    
/usr/bin:/usr/bin:/home/pierre/.cargo/bin:/data0/opt/miniconda3/condabin/app:/data0/opt/miniconda3/condabin:/home/pierre/.pyenv/shims:/home/pierre/.pyenv/bin:/home/pierre/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
```
For the second command (path from Python), it seems that pyenv adds some paths.

**Comment by laloch** [comment]
I see. So it seems Xonsh is not guilty, right?

**Comment by paugier** [comment]
I don't know who's guilty but it is simpler without pyenv and there is still something strange:
```
$ echo $PATH                                                                                                  
/home/pierre/.cargo/bin:/data0/opt/miniconda3/condabin/app:/data0/opt/miniconda3/condabin:/home/pierre/.pyenv/shims:/home/pierre/.pyenv/bin:/home/pierre/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
$ /usr/bin/python3 -c "import os; print(os.environ['PATH'])"                                                  
/home/pierre/.cargo/bin:/data0/opt/miniconda3/condabin/app:/data0/opt/miniconda3/condabin:/home/pierre/.pyenv/shims:/home/pierre/.pyenv/bin:/home/pierre/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
$ mpirun -np 1 /usr/bin/python3 -c "import os; print(os.environ['PATH'])"                                     
/usr/bin:/usr/bin:/home/pierre/.cargo/bin:/data0/opt/miniconda3/condabin/app:/data0/opt/miniconda3/condabin:/home/pierre/.pyenv/shims:/home/pierre/.pyenv/bin:/home/pierre/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
$ mpirun -x PATH -np 1 /usr/bin/python3 -c "import os; print(os.environ['PATH'])"                             
/usr/bin:/home/pierre/.cargo/bin:/data0/opt/miniconda3/condabin/app:/data0/opt/miniconda3/condabin:/home/pierre/.pyenv/shims:/home/pierre/.pyenv/bin:/home/pierre/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
```

**Comment by laloch** [comment]
Fascinating :)
So `mpirun` insists on prepending the PATH with at least one "/usr/bin". Funny.
What about `mpirun --path "$PATH" ...`?

**Comment by paugier** [comment]
`--path` does not help. And it also works with a much simpler `PATH`:

```
$ $PATH = "/usr/bin:/sbin:/bin"                                                                               
$ /usr/bin/python3 -c "import os; print(os.environ['PATH'])"                                                  
/usr/bin:/sbin:/bin
$ mpirun -np 1 /usr/bin/python3 -c "import os; print(os.environ['PATH'])"                                     
/usr/bin:/usr/bin:/usr/bin:/sbin:/bin
$ mpirun --path "$PATH" -np 1 /usr/bin/python3 -c "import os; print(os.environ['PATH'])"                      
/usr/bin:/usr/bin:/usr/bin:/sbin:/bin
$ mpirun -x PATH -np 1 /usr/bin/python3 -c "import os; print(os.environ['PATH'])"                             
/usr/bin:/usr/bin:/sbin:/bin
```

**Comment by jonaslb** [comment]
I'm seeing this odd thing in Fedora as well.

```
$ echo $PATH                                                                                                  
/usr/lib64/openmpi/bin:/home/jonas/.local/share/cargo/bin:/home/jonas/.local/bin:/home/jonas/bin:/home/jonas/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin
$ python3 -c "import os; print(os.environ['PATH'])"                                                           
/usr/lib64/openmpi/bin:/home/jonas/.local/share/cargo/bin:/home/jonas/.local/bin:/home/jonas/bin:/home/jonas/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin
$ bash -c """mpirun -n 1 python3 -c "import os; print(os.environ['PATH'])" """                                
/usr/lib64/openmpi/bin:/home/jonas/.local/share/cargo/bin:/home/jonas/.local/bin:/home/jonas/bin:/home/jonas/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin
$ mpirun -n 1 bash -c "echo \$PATH"                                                                           
/usr/lib64/openmpi/bin:/home/jonas/.local/share/cargo/bin:/home/jonas/.local/bin:/home/jonas/bin:/home/jonas/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin
$ mpirun -n 1 python3 -c "import os; print(os.environ['PATH'])"                                               
/usr/lib64/openmpi/bin:/usr/lib64/openmpi/bin:/usr/lib64/openmpi/bin:/home/jonas/.local/share/cargo/bin:/home/jonas/.local/bin:/home/jonas/bin:/home/jonas/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin
```

In my case it prepends PATH with /usr/lib64/openmpi/bin *twice*. And only for `mpirun python`, not for `mpirun bash` or `bash mpirun python`.

**Comment by laloch** [comment]
Alas, I don't see how we at Xonsh could help you. I think you guys should file a bug report at Open MPI issue tracker.

**Comment by paugier** [comment]
Isn't it a xonsh issue that OpenMPI doesn't feel the same when it is called from xonsh than when it is called from bash or fish? The shell should not influence the result of a program, no?
```
pierre@kth ~ $ cat print_path.sh                                                
#!/usr/bin/env bash
echo $PATH
pierre@kth ~ $ $PATH = "/usr/bin:/sbin:/bin"                                    
pierre@kth ~ $ ./print_path.sh                                                  
/usr/bin:/sbin:/bin
pierre@kth ~ $ mpirun -np 1 ./print_path.sh                                     
/usr/bin:/usr/bin:/usr/bin:/sbin:/bin
pierre@kth ~ $ bash                                                             
pierre@kth:~$ export PATH="/usr/bin:/sbin:/bin"
pierre@kth:~$ mpirun -np 1 ./print_path.sh
/usr/bin:/sbin:/bin
pierre@kth:~$ fish
Welcome to fish, the friendly interactive shell
pierre@kth ~> export PATH="/usr/bin:/sbin:/bin"
pierre@kth ~> mpirun -np 1 ./print_path.sh
/usr/bin:/sbin:/bin
```

**Comment by laloch** [comment]
> The shell should not influence the result of a program, no?

Yes, of course. Unless, of course, the "program" does actively something to break its own environment in dependence on the parent process, process group, controlling terminal, moon phase or other arbitrary condition we cannot assume.

**Comment by paugier** [comment]
Yes, I see what you mean. And it seems to be the case!

A simple workaround, use `env` before `mpirun`
```
$ mpirun -np 1 ./print_path.sh                                     
/usr/bin:/usr/bin:/usr/bin:/sbin:/bin
$ env mpirun -np 1 ./print_path.sh                                 
/usr/bin:/sbin:/bin
```
Thank you for your time @laloch

**Comment by ggouaillardet** [comment]
@laloch 

Under the hood,
`mpirun -np 1 ./print_path.sh
will result in (in `bash`)
`execve ("/usr/bin/mpirun", ["mpirun", "-np", "1", "./print_path.sh"], [...])`
but in `xonsh`
`execve ("/usr/bin/mpirun", ["/usr/bin/mpirun", "-np", "1", "./print_path.sh"], [...])`

Open MPI will then "do stuff" depending on whether `argv[0]` is `mpirun` or `/usr/bin/mpirun` (and no other esoteric conditions :-))

Bottom line, is `xonsh` does not pass to `execve()` the command provided by the user, but its full path instead. Is this a bug or a feature? I think the former, but this is clearly not my call.

**Comment by laloch** [comment]
@ggouaillardet, thanks for investigating the issue.
Looking at the `execve` documentation I'd say you are right - It's a bug ... in `bash`. But if this misuse of `execve` arguments is a common practice, we can adjust. Unless it's a feature of Python libraries.

**Comment by ggouaillardet** [comment]
FWIW, `man execve` on OSX says

> The argument argv is a pointer to a null-terminated array of character pointers to null-terminated character strings.  [...] by custom, the first element should be the name of the executed program (for example, the last component of path).

My understanding is that the "last component" of `/usr/bin/mpirun` is `mpirun`, so not only `bash` is correct but it also honors the custom.

**Comment by laloch** [comment]
Interesting. Linux (man 2 execve) documentation says:
> By convention, the first of these strings (i.e., argv[0]) should contain the filename associated with the file being executed.

"filename" for me is a name used to **uniquely** idetify a file in a file system.

**Comment by ggouaillardet** [comment]
To me, `filename` is very ambiguous and does not suggest "name of the file" (e.g. `mpirun`) or the full path (and btw, with the use of symbolic links, several paths can point to the very same inode (`mpirun` here)).

**Comment by laloch** [comment]
So @ggouaillardet, you basically say that the following commands would lead to execution in different environments, right? That sounds really weird.
```
$ which mpirun
/usr/bin/mpirun
$ cd /usr
$ mpirun
$ bin/mpirun
$ /usr/bin/mpirun
```

**Comment by ggouaillardet** [comment]
under bash and with a default install, `mpirun` and `/usr/bin/mpirun` will indeed have the app executed with a different `$PATH`.
`mpirun` and `bin/mpirun` will lead to the same environment w.r.t. `$PATH`

**Comment by paugier** [comment]
The Linux documentation uses both `pathname` (first argument of `execve`) and `filename` (content of `argv[0]`).

I also think it is weird to modify PATH depending on how the program is called. But I guess it is useful in some cases for OpenMPI.

Anyway, in terms of how the programs should be called, shouldn't xonsh have to follow bash even if bash is wrong?

**Comment by laloch** [comment]
> Anyway, in terms of how the programs should be called, shouldn't xonsh have to follow bash even if bash is wrong?

Yes, definitely. Regardless of what I think about the issue, Xonsh should definitely behave consistently with sh/bash. PR incoming.
